Allow for lazy evaluation of Jinja2 expressions

mkrizek opened the pull request:
##### SUMMARY

This will allow for:

```yaml
- hosts: localhost
  gather_facts: no
  vars:
    foo: "{{ nested_undefined_variable }}"
  tasks:
    - debug:
      when: True or foo
```

Now, the conditional is evaluated to `True`. Before, this would fail on `foo` being undefined although it wasn't necessary to evaluate it.

Also, this changes behavior(!):

```yaml
- hosts: localhost
  gather_facts: no
  vars:
    foo: "{{ nested_undefined_variable }}"
  tasks:
    - assert:
        that:
          - foo is undefined
```

Before:
```
TASK [assert] *****************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "The conditional check 'foo is undefined' failed. The error was: error while evaluating conditional (foo is undefined): 'nested_undefined_variable' is undefined"}
```

Now:
```
TASK [assert] *******************************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}
```

Fixes #56017

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/template/vars.py`

mkrizek committed:
Allow for lazy evaluation of Jinja2 expressions

ci_complete

bcoca commented on test/units/playbook/test_conditional.py:
```diff
@@ -95,10 +95,8 @@ def test_dict_undefined_values(self):
-                                "The conditional check 'some_defined_dict_with_undefined_values is defined' failed.",
-                                self._eval_con,
-                                when, variables)
+        ret = self._eval_con(when, variables)
```
this should not change

giner commented:
Any plans to move this forward?

mkrizek committed:
Clean up

sivel commented:
/azp run

sivel commented:
I wanted to outline the lingering unit test issue, that is asserting a behavior that I am unsure we should keep.

The unit test is represented by this playbook:

```yaml
- hosts: localhost
  gather_facts: false
  tasks:
    - assert:
        that:
          - some_defined_dict_with_undefined_values is defined
      vars:
        dict_value: 1
        some_defined_dict_with_undefined_values:
          key1: value1
          key2: '{{ dict_value }}'
          key3: '{{ undefined_dict_value }}'
```

Currently using devel results in the following:

```
TASK [assert] ********************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "The conditional check 'some_defined_dict_with_undefined_values is defined' failed. The error was: error while evaluating conditional (some_defined_dict_with_undefined_values is defined): {'key1': 'value1', 'key2': '{{ dict_value }}', 'key3': '{{ undefined_dict_value }}'}: 'undefined_dict_value' is undefined"}
```

With this PR:

```
TASK [assert] ********************************************************************************************************
fatal: [localhost]: FAILED! => {
    "assertion": "some_defined_dict_with_undefined_values is defined",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

To summarize, the current behavior is to raise an `AnsibleError` due to `undefined_dict_value` being undefined.  With the change in this PR, that does not happen, and instead `some_defined_dict_with_undefined_values` is evaluated as being undefined, because one of the values is undefined.

nitzmahone commented:
I'm honestly not sure either of them is correct; I assume this is because of our eager recursive templating of vars... If nothing "touches" an undefined value in the dictionary, do we care? I could see it being problematic for things like `debug` or anything else that tries to naively render the dict.

Regardless, assuming we keep the present behavior in some form (where this becomes an error at task template-time), the loss of context on the error trip is unfortunate. I'm betting we can do something about that- if not in the current codebase, data tagging keeps a recursive template context for other purposes that we can probably also unwind for a recursive undefined to keep the context around the path to the actual undefined value that's causing the error.

mkrizek committed:
Remove test, we are changing behavior

mkrizek commented:
@nitzmahone I am not sure what you mean by the loss of context, but if used directly it still fails with the error message providing the same information as it currently does in devel:
```yaml
- hosts: localhost
  gather_facts: false
  tasks:
    - command: "echo {{ some_defined_dict_with_undefined_values }}"
      vars:
        dict_value: 1
        some_defined_dict_with_undefined_values:
          key1: value1
          key2: '{{ dict_value }}'
          key3: '{{ undefined_dict_value }}'
```

```
TASK [command] **************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"msg": "The task includes an option with an undefined variable. The error was: {'key1': 'value1', 'key2': '{{ dict_value }}', 'key3': '{{ undefined_dict_value }}'}: 'undefined_dict_value' is undefined\n\nThe error appears to be in '/Users/mkrizek/src/ansible/test2.yml': line 4, column 7, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  tasks:\n    - command: \"echo {{ some_defined_dict_with_undefined_values }}\"\n      ^ here\nWe could be wrong, but this one looks like it might be an issue with\nmissing quotes. Always quote template expression brackets when they\nstart a value. For instance:\n\n    with_items:\n      - {{ foo }}\n\nShould be written as:\n\n    with_items:\n      - \"{{ foo }}\"\n"}
```

mkrizek committed:
Change to a breaking change

mkrizek committed:
Porting guide

mkrizek committed:
ci_complete

mkrizek commented:
I noted in the changelog that this is a breaking change and removed the failing test.

I also added a porting guide for 2.13 but I understand that it would be wiser to postpone this to 2.14 (if we decide to merge this at all of course) :-)

nitzmahone commented:
Oh, I was just going off sivel's test output above, which at least in that case seems to provide a lot less information after this change than before. That may be because we're wrapping it as a conditional, in which case there's probably something more we could do to preserve the extra information about the underlying problem.

mkrizek committed:
Move to 2.14

mkrizek committed:
ci_complete

sivel commented:
/azp run
